From a304902564b3f27080da30c0e5c9adfe6f1071c0 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C3=8D=C3=B1igo=20Huguet?= <ihuguet@riseup.net>
Date: Thu, 16 Oct 2025 12:45:41 +0200
Subject: [PATCH] import/export: add special comment for nm-auto-defaults
 detection

In commit 3ea80883fef we added the nm-auto-defaults property to allow
interpreting configurations just like Libreswan, without assumptions
like setting leftmodecfgclient=yes by default, which is different to
what Libreswan would do.

This means that with nm-auto-defaults=no the behaviour is slightly
different. When we export a connection with nm-auto-defaults=no, we
cannot reflect it in the exported file, as it is not a valid option in
Libreswan. If we import back the same file, it will be imported as
nm-auto-defaults=yes (by default), thus it will have different
behaviour. This is wrong, export & import should be symetric.

Make the "write" function to emit a `# nm-auto-defaults=no` comment in
the exported file. When importing, make the "parse" function to
interpret this special comment as `nm-auto-defaults=no`. This will
ensure that we can export & import symetrically.

The comment can also be added manually to any file that users want to
import.

Note: increase the minimum GLib version to 2.44 to use g_autoptr. It was
released more than 10 years ago, so we're quite safe.
---
 configure.ac        |  6 ++---
 shared/test-utils.c | 57 ++++++++++++++++++++++++++++++++++++++++++---
 shared/utils.c      | 25 ++++++++++++++++++--
 3 files changed, 80 insertions(+), 8 deletions(-)

diff --git a/configure.ac b/configure.ac
index 9b9677a..f87ffdf 100644
--- a/configure.ac
+++ b/configure.ac
@@ -116,9 +116,9 @@ AC_DEFINE_UNQUOTED(GETTEXT_PACKAGE,"$GETTEXT_PACKAGE", [Gettext package])
 IT_PROG_INTLTOOL([0.35])
 AM_GLIB_GNU_GETTEXT
 
-PKG_CHECK_MODULES(GLIB, gio-unix-2.0 >= 2.36)
-GLIB_CFLAGS="$GLIB_CFLAGS -DGLIB_VERSION_MIN_REQUIRED=GLIB_VERSION_2_36"
-GLIB_CFLAGS="$GLIB_CFLAGS -DGLIB_VERSION_MAX_ALLOWED=GLIB_VERSION_2_36"
+PKG_CHECK_MODULES(GLIB, gio-unix-2.0 >= 2.44)
+GLIB_CFLAGS="$GLIB_CFLAGS -DGLIB_VERSION_MIN_REQUIRED=GLIB_VERSION_2_44"
+GLIB_CFLAGS="$GLIB_CFLAGS -DGLIB_VERSION_MAX_ALLOWED=GLIB_VERSION_2_44"
 
 PKG_CHECK_MODULES(LIBNL, libnl-3.0 >= 3.2.8)
 
diff --git a/shared/test-utils.c b/shared/test-utils.c
index 8bf5888..c523c9c 100644
--- a/shared/test-utils.c
+++ b/shared/test-utils.c
@@ -183,7 +183,7 @@ test_config_write (void)
 	nm_setting_vpn_add_data_item (s_vpn, "leftrsasigkey", "hello");
 	nm_setting_vpn_add_data_item (s_vpn, "rightrsasigkey", "world");
 	nm_setting_vpn_add_data_item (s_vpn, "right", "11.12.13.14");
-	nm_setting_vpn_add_data_item (s_vpn, "nm-auto-defaults", "false");
+	nm_setting_vpn_add_data_item (s_vpn, "nm-auto-defaults", "no");
 	nm_setting_vpn_add_data_item (s_vpn, "leftsendcert", "always");
 	nm_setting_vpn_add_data_item (s_vpn, "rightca", "%same");
 	s_vpn_sanitized = sanitize_setting_vpn (s_vpn, &error);
@@ -191,6 +191,9 @@ test_config_write (void)
 	str = nm_libreswan_get_ipsec_conf (4, s_vpn_sanitized, "conn", NULL, FALSE, TRUE, &error);
 	g_assert_no_error (error);
 	g_assert_cmpstr (str, ==,
+                         "# NetworkManager specific configs, don't remove:\n"
+                         "# nm-auto-defaults=no\n"
+                         "\n"
                          "conn conn\n"
                          " ikev2=insist\n"
                          " right=11.12.13.14\n"
@@ -294,7 +297,7 @@ test_config_write (void)
 	g_object_unref (s_vpn);
 
 	s_vpn = NM_SETTING_VPN (nm_setting_vpn_new ());
-	nm_setting_vpn_add_data_item (s_vpn, "nm-auto-defaults", "false");
+	nm_setting_vpn_add_data_item (s_vpn, "nm-auto-defaults", "no");
 	nm_setting_vpn_add_data_item (s_vpn, "rightcert", "\"cert\"");
 	s_vpn_sanitized = sanitize_setting_vpn (s_vpn, &error);
 	g_assert_error (error, NM_UTILS_ERROR, NM_UTILS_ERROR_INVALID_ARGUMENT);
@@ -303,7 +306,7 @@ test_config_write (void)
 	g_object_unref (s_vpn);
 
 	s_vpn = NM_SETTING_VPN (nm_setting_vpn_new ());
-	nm_setting_vpn_add_data_item (s_vpn, "nm-auto-defaults", "false");
+	nm_setting_vpn_add_data_item (s_vpn, "nm-auto-defaults", "no");
 	s_vpn_sanitized = sanitize_setting_vpn (s_vpn, &error);
 	g_assert_error (error, NM_UTILS_ERROR, NM_UTILS_ERROR_INVALID_ARGUMENT);
 	g_assert_null (s_vpn_sanitized);
@@ -612,6 +615,54 @@ test_config_read (void)
 	g_object_unref (s_vpn);
 	g_clear_pointer (&con_name, g_free);
 
+	/* With the '# nm-auto-defaults=no' special comment */
+	s_vpn = nm_libreswan_parse_ipsec_conf (
+		"# nm-auto-defaults=no\n"
+		"conn conn\n"
+		" ikev2=insist\n"
+		" right=11.12.13.14\n"
+		" rightrsasigkey=\"world\"\n"
+		" leftrsasigkey=\"hello\"\n"
+		" leftsendcert=always\n"
+		" rightca=\"%same\"\n",
+		&con_name,
+		&error);
+	g_assert_no_error (error);
+	g_assert_cmpstr (nm_setting_vpn_get_data_item (s_vpn, "ikev2"), ==, "insist");
+	g_assert_cmpstr (nm_setting_vpn_get_data_item (s_vpn, "leftrsasigkey"), == , "hello");
+	g_assert_cmpstr (nm_setting_vpn_get_data_item (s_vpn, "rightrsasigkey"), == , "world");
+	g_assert_cmpstr (nm_setting_vpn_get_data_item (s_vpn, "right"), == , "11.12.13.14");
+	g_assert_cmpstr (nm_setting_vpn_get_data_item (s_vpn, "nm-auto-defaults"), == , "no");
+	g_assert_cmpstr (nm_setting_vpn_get_data_item (s_vpn, "leftsendcert"), == , "always");
+	g_assert_cmpstr (nm_setting_vpn_get_data_item (s_vpn, "rightca"), == , "%same");
+	g_assert_cmpstr (nm_setting_vpn_get_data_item (s_vpn, "rightca"), == , "%same");
+	g_object_unref (s_vpn);
+	g_clear_pointer (&con_name, g_free);
+
+	/* With the '# nm-auto-defaults=no' special comment, different spacing */
+	s_vpn = nm_libreswan_parse_ipsec_conf (
+		"#nm-auto-defaults  	= 	 no  	 \n"
+		"conn conn\n"
+		" ikev2=insist\n"
+		" right=11.12.13.14\n"
+		" rightrsasigkey=\"world\"\n"
+		" leftrsasigkey=\"hello\"\n"
+		" leftsendcert=always\n"
+		" rightca=\"%same\"\n",
+		&con_name,
+		&error);
+	g_assert_no_error (error);
+	g_assert_cmpstr (nm_setting_vpn_get_data_item (s_vpn, "ikev2"), ==, "insist");
+	g_assert_cmpstr (nm_setting_vpn_get_data_item (s_vpn, "leftrsasigkey"), == , "hello");
+	g_assert_cmpstr (nm_setting_vpn_get_data_item (s_vpn, "rightrsasigkey"), == , "world");
+	g_assert_cmpstr (nm_setting_vpn_get_data_item (s_vpn, "right"), == , "11.12.13.14");
+	g_assert_cmpstr (nm_setting_vpn_get_data_item (s_vpn, "nm-auto-defaults"), == , "no");
+	g_assert_cmpstr (nm_setting_vpn_get_data_item (s_vpn, "leftsendcert"), == , "always");
+	g_assert_cmpstr (nm_setting_vpn_get_data_item (s_vpn, "rightca"), == , "%same");
+	g_assert_cmpstr (nm_setting_vpn_get_data_item (s_vpn, "rightca"), == , "%same");
+	g_object_unref (s_vpn);
+	g_clear_pointer (&con_name, g_free);
+
 	s_vpn = nm_libreswan_parse_ipsec_conf (
 		"conn my_con\n",
 		&con_name,
diff --git a/shared/utils.c b/shared/utils.c
index 9cb36be..c188e5b 100644
--- a/shared/utils.c
+++ b/shared/utils.c
@@ -482,6 +482,7 @@ nm_libreswan_get_ipsec_conf (int ipsec_version,
                              GError **error)
 {
 	nm_auto_free_gstring GString *ipsec_conf = NULL;
+	gboolean auto_defaults;
 	const char *val;
 	int i;
 
@@ -493,6 +494,16 @@ nm_libreswan_get_ipsec_conf (int ipsec_version,
 		return NULL;
 
 	ipsec_conf = g_string_sized_new (1024);
+
+	auto_defaults = _nm_utils_ascii_str_to_bool (
+		nm_setting_vpn_get_data_item (s_vpn_sanitized, NM_LIBRESWAN_KEY_NM_AUTO_DEFAULTS),
+		TRUE);
+	if (!auto_defaults) {
+		g_string_append(ipsec_conf, "# NetworkManager specific configs, don't remove:\n");
+		g_string_append(ipsec_conf, "# nm-auto-defaults=no\n");
+		g_string_append(ipsec_conf, "\n");
+	}
+
 	g_string_append_printf (ipsec_conf, "conn %s\n", con_name);
 
 	for (i = 0; params[i].name != NULL; i++) {
@@ -573,6 +584,8 @@ static const char line_match[] =
 	")?"					/* (or just blank line) */
 	"\\s*(?:#.*)?$";			/* optional comment */
 
+static const char no_auto_match[] = "#\\s*nm-auto-defaults\\s*=\\s*no";
+
 NMSettingVpn *
 nm_libreswan_parse_ipsec_conf (const char *ipsec_conf,
                                char **out_con_name,
@@ -584,7 +597,8 @@ nm_libreswan_parse_ipsec_conf (const char *ipsec_conf,
 	gs_free char *con_name = NULL;
 	GMatchInfo *match_info = NULL;
 	GError *parse_error = NULL;
-	GRegex *line_regex;
+	g_autoptr(GRegex) line_regex = NULL;
+	g_autoptr(GRegex) no_auto_regex = NULL;
 	const char *old, *new;
 	const char *rekey;
 	char *key, *val;
@@ -596,6 +610,8 @@ nm_libreswan_parse_ipsec_conf (const char *ipsec_conf,
 
 	line_regex = g_regex_new (line_match, G_REGEX_RAW, 0, NULL);
 	g_return_val_if_fail (line_regex, NULL);
+	no_auto_regex = g_regex_new (no_auto_match, G_REGEX_RAW, 0, NULL);
+	g_return_val_if_fail (no_auto_regex, NULL);
 
 	s_vpn = NM_SETTING_VPN (nm_setting_vpn_new ());
 
@@ -611,6 +627,11 @@ nm_libreswan_parse_ipsec_conf (const char *ipsec_conf,
 			break;
 		}
 
+		if (g_regex_match(no_auto_regex, lines[i], 0, NULL)) {
+			nm_setting_vpn_add_data_item(s_vpn, NM_LIBRESWAN_KEY_NM_AUTO_DEFAULTS, "no");
+			continue;
+		}
+
 		key = g_match_info_fetch (match_info, 1); /* Key */
 		val = g_match_info_fetch (match_info, 2); /* Unquoted value */
 		/* Without fix from
@@ -666,7 +687,7 @@ nm_libreswan_parse_ipsec_conf (const char *ipsec_conf,
 		if (parse_error)
 			break;
 	}
-	g_regex_unref (line_regex);
+
 	if (parse_error) {
 		g_propagate_error (error, parse_error);
 		return NULL;
-- 
2.51.0

